Code refactoring

dhruv10 opened the pull request:
1. Followed eslint-airbnb guidelines
- constants file, helper function file, mutability constraints
2. Changed the directory structure

dhruv10 committed:
Setup eslint-airbnb and prettier along with config
Done some linting fixes in Home

dhruv10 committed:
Removed GitHusky

dhruv10 committed:
temporarily disbled class-method-use-this

dhruv10 committed:
extraneous dependancy issue fixed

dhruv10 committed:
eslintrc no-extraneous-depencency removed

dhruv10 committed:
1. Changed directory-structure for vizone
2. Created vizone/utils folder
3. Added a constants file
4. Added a file for helper functions

dhruv10 committed:
Code refactored
1. Environment token for Mapbox
2. ESlint airbnb guidelines
3. Changed directory structure

dhruv10 committed:
changes in eslintrc rule - removed jsx spreading

dhruv10 committed:
removed unnecessary files

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+  }
+}
+
+export default App;
```
The component should be named VizOne or something of that sort.

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+            onHover: hover => this.onHover(hover),
+            settings
+          })}
+          effeccontrol-panelts={[CONSTANTS.lightingEffect]}
```
`effeccontrol-panelts` Is there a typo error here or the correct API is being used?

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+              </div>
+            </div>
+          </div>
+        )}
```
This should be moved to a functional component and it's reference being used in the return of the render.

prastut replied:
```
style={{
              transform: `translate(${hover.x}px, ${hover.y}px)`
            }}

```

No inline styles. 


For CSS-in-JS we would need to move to styled-components or my favorite JSS

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+            data: points,
+            onHover: hover => this.onHover(hover),
+            settings
+          })}
```
This should be moved out of the return function and set here https://github.com/prastut/foamviz/pull/45/files#diff-11793eaf7aaec82bbd4bfae5d53fa5d0R139

Only a variable reference to the return value of the `renderLayers` call should be referenced here.

dhruv10 replied:
The link is not working. Please re-share

prastut commented on src/viz-one/utils/constants.js:
```diff
@@ -0,0 +1,102 @@
+    min: 80,
+    max: 100
+  }
+};
```
Lighting and controls are separate concerns. That should be in separate files.

prastut commented on src/viz-one/utils/helper.js:
```diff
@@ -0,0 +1,26 @@
+    sum += item.stakedvalue;
+  });
+  return sum.toFixed(2);
+}
```
Use `reduce` to sum in a functional manner. https://stackoverflow.com/questions/1230233/how-to-find-the-sum-of-an-array-of-numbers

And see the beauty of functional code :D it makes code look like poetry to me.

prastut commented on src/viz-one/utils/helper.js:
```diff
@@ -0,0 +1,26 @@
+    .then(res => res.json())
+    .then(json => {
+      return Promise.resolve(json.USDC_BTC.last * json.BTC_FOAM.last);
+    });
```
Let's move to https://github.com/axios/axios for API calls.

prastut commented on src/viz-one/controls.js:
```diff
@@ -0,0 +1,153 @@
+import * as CONSTANTS from './utils/constants';
+import './index.css';
+
+function _onValueChange(settingName, newValue, props) {
```
Please follow the functions styleguide: https://github.com/airbnb/javascript#functions

prastut commented on src/viz-one/controls.js:
```diff
@@ -0,0 +1,153 @@
+      </div>
+    </div>
+  );
+};
```
This is a separate functional component (file)

prastut commented on src/viz-one/controls.js:
```diff
@@ -0,0 +1,153 @@
+      </div>
+    </div>
+  );
+};
```
This is a separate functional component file.

prastut commented on src/viz-one/controls.js:
```diff
@@ -0,0 +1,153 @@
+  );
+};
+
+export default LayerControls;
```
Component Names should be after ideally what they do. Currently the problem with this: 
1. Whose Viz are these layercontrols for? 
2. What does LayerControls mean? Does it control deck.gl layers? 

Ideally the file should be named VizOneControlPanel

prastut replied:
Todo: 
A base filename should exactly match the name of its default export: https://github.com/airbnb/javascript#naming--filename-matches-export
LayerControls != controls.js

prastut commented on src/viz-one/deckgl-layers.js:
```diff
@@ -0,0 +1,48 @@
+import { HexagonLayer } from 'deck.gl';
+import * as CONSTANTS from './utils/constants';
+
+function renderLayers(props) {
```
https://github.com/airbnb/javascript#functions

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+        lat: '40.636102'
+      }
+    };
+    this.fetchData(currBbox);
```
These defaults should be placed in external scope to the core class, not internal. 

Defining it outside the scope with caps to inform the programmer that this is a default value being passed to the component.

prastut replied:
Search `How to define propTypes, defaultProps, contextTypes, etc...` in https://github.com/airbnb/javascript/tree/master/react#ordering

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+        lat: '40.636102'
+      }
+    };
+    this.fetchData(currBbox);
```
Problems with this: 

1. Currently, data is being fetched for NewYork first irrespective of whether we get user information for not. 
2. There are 3 setStates happening across these 3 functions. State Updates are async in React, and there is no deterministic way of defining what will occur step by step in this (read: https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous)

prastut commented on src/viz-one/deckgl-layers.js:
```diff
@@ -0,0 +1,48 @@
+        onHover,
+        ...settings
+      })
+  ];
```
This is returning an array? Is that correct? Todo: discuss

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+    };
+  }
+
+  async componentDidMount() {
```
componentDidMount() is placed just before render function to improve readability.

dhruv10 replied:
Eslint react plugin says custom methods should occur before lifecycle methods https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/sort-comp.md

prastut replied:
@dhruv10 triple check, wait for 1 minute after you have written and then submit. 

Eslint react plugin says custom methods should occur ~before~ **after** lifecycle methods. 

---

Sure, open to new things. Let's try what eslint says.

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+    });
+  }
+
+  async onHover({ x, y, object }) {
```
What is x,y, object? 

Improve readability of code. I think x,y are screen-coordinates and object contains latLong. 


object -> pointHoveredObject

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+        sumOfFoamTokens: getSumOfFoamTokens(object.points),
+        sumValInUSD: (
+          getSumOfFoamTokens(object.points) * FOAMTokenInUSD
+        ).toFixed(2)
```
Repeated calls to getSumOfFoamTokens. Can be called once and used twice.

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+      };
+      this.setState({ hover: { x, y, hoveredObject: object, details } });
+    } else {
+      this.setState({ hover: { x, y, hoveredObject: object } });
```
Discuss else clause. If there is no object, why are we setting state to NaN? Why are we updating state at all?

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+    this.setState({ settings });
+  }
+
+  fetchData(bbox) {
```
fetchData -> fetchAllPOIDetailsInCurrentViewport

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+      `https://map-api-direct.foam.space/poi/filtered?swLng=${bbox._sw.lng}&swLat=${bbox._sw.lat}&neLng=${bbox._ne.lng}&neLat=${bbox._ne.lat}&limit=10000&offset=0`
+    )
+      .then(result => result.json())
+      .then(json => {
```
json => arrayOfPOIObjects

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+      .then(result => result.json())
+      .then(json => {
+        const points = [];
+        json.forEach((item, index) => {
```
forEach to be ported to map, need not create external points array. 

https://codeburst.io/javascript-map-vs-foreach-f38111822c0f

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+              parseFloat(pointCoords[0].toFixed(4)),
+              parseFloat(pointCoords[1].toFixed(4))
+            ],
+            pickup: item[2],
```
What is pickup?

prastut commented on src/viz-one/index.js:
```diff
@@ -0,0 +1,194 @@
+          points
+        });
+      });
+  }
```
I can't seem to find the code that checks for data sanity i.e these cases can happen when viewport changes and data is fetched: 
1. previous state === data fetched i.e same data has been fetched i.e there is no new data in the current viewport hence state shouldn't update. Incremental data change. 
2. There is overlap between the previous state and new data fetched, i.e there is some new data that is fetched and only the new data should be appended to the previous state. 
3. No overlap between the new data and previous state === entirely new data is being fetched, rerendering only cd

prastut replied:
I think using the solution mentioned in `is the data change incremental?` will solve the problem that we are facing. https://github.com/uber/deck.gl/blob/master/docs/developer-guide/performance.md

dhruv10 committed:
added linting script

dhruv10 committed:
linting errors fixed in vizone/index.js

dhruv10 committed:
WIP ...requested changes in code refactoring

dhruv10 committed:
wip ...code refactoring

prastut commented on src/VizPOIAnalytics/utils/layerProperties.js:
```diff
@@ -0,0 +1,43 @@
+  specularColor: [51, 51, 51],
+});
+
+const LIGHT_SETTINGS = {
```
Why is this part of layerProperties.js file?

prastut commented on src/VizPOIAnalytics/POIAnalyticsControlPanel.js:
```diff
@@ -0,0 +1,148 @@
+  }
+}
+
+function POIAnalyticsControlPanel(props) {
```
```javascript
    // bad
    function POIAnalyticsControlPanel() {
      // ...
    }

    // good
   const POIAnalyticsControlPanel = (props) => { }
    ```

prastut replied:
Please apply this to all function calls

dhruv10 replied:
![image](https://user-images.githubusercontent.com/32517802/68523981-6e086900-02e6-11ea-9332-1fa189e02e24.png)
Followed this only but it seems they have multiple good conventions. Shifting to what you're suggesting. Earlier I got confused between the two. Even, I disliked the long lexical name convention

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+    };
+  }
+
+  async componentDidMount() {
```
async without await

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+  async onHover({ x, y, object }) {
+    const { FOAMTokenInUSD } = this.state;
+    if (object) {
+      const sumOfFoamTokens = object.points
```
Should be moved into a function

```
const getToolTipFormattedDetails  = (object) => {}

this.setState(..... details: getToolTipFormattedDetails(object)
```

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+          [key]: CONSTANTS.HEXAGON_CONTROLS[key].value,
+        }),
+        {},
+      ),
```
Needs to be made simpler

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+          hoveredObject: object,
+          details,
+        },
+      });
```
Should be moved into a function - which is not part of this class scope. 

```
const getToolTipFormattedDetails  = (object) => {}

this.setState(..... details: getToolTipFormattedDetails(object)
```

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+      },
+    };
+    this.fetchAllPOIDetailsInCurrentViewport(boundingBox);
+  }
```
So there are three different functions here: 

1. getBoundingBoxDetailsFromCurrentViewport which fetches the current viewPort. 
2. fetchPOIDetailsFromFOAMAPI which given bounding box fetches it from FOAMAPI
3. fetchAllPOIDetailsInCurrentViewport which encapsulates functionality from 1 and 2

Currently the code is confusing as it uses both getDataForCurrentViewport & fetchAllPOIDetailsInCurrentViewport which are actually just a single function. 


```
async fetchAllPOIDetailsInCurrentViewport() {

const boundingBoxDetailsFromCurrentViewPort = getBoundingBoxDetailsFromCurrentViewport(this.mapRef.getMap().getBounds())

const  pointsInCurrentViewPort = await fetchPOIDetailsFromFOAMAPI(boundingBoxDetailsFromCurrentViewPort)

this.setState({points: pointsInCurrentViewPort})
}

```

Remember this: One function must **do** only one thing. In the above `fetchAllPOIDetailsInCurrentViewport` the doing part is only setting state, rest is getting data for setting state.

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+            ],
+            stakedvalue: temp,
+          };
+        });
```
Moved to a function call

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+            }}
+            mapStyle={CONSTANTS.MAP_STYLE}
+            mapboxApiAccessToken={process.env.REACT_APP_MAPBOX_ACCESS_TOKEN}
+            onLoad={this.getDataForCurrentViewport}
```
fetchAllPOIDetailsInCurrentViewport

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+          effects={[lightingEffect]}
+          initialViewState={{ ...viewport }}
+          controller
+          onDragEnd={this.getDataForCurrentViewport}
```
fetchAllPOIDetailsInCurrentViewport

prastut commented on src/VizPOIAnalytics/utils/helper.js:
```diff
@@ -0,0 +1,21 @@
+  const sum = points.reduce((prevvalue, item) => prevvalue + item.stakedvalue, 0).toFixed(2);
+  return sum;
+}
+
```
```
export const getSumOfFoamTokens = points =>
  points
    .reduce((prevvalue, item) => prevvalue + item.stakedvalue, 0)
    .toFixed(2);
```

prastut commented on src/VizPOIAnalytics/utils/layerProperties.js:
```diff
@@ -0,0 +1,43 @@
+  specularColor: [51, 51, 51],
+});
+
+const LIGHT_SETTINGS = {
+  lightsPosition: [-73.8, 40.5, 8000, -74.2, 40.9, 8000],
+  ambientRatio: 0.4,
+  diffuseRatio: 0.6,
+  specularRatio: 0.2,
+  lightsStrength: [0.8, 0.0, 0.8, 0.0],
+  numberOfLights: 2,
+};
```
Why is this part of layerProperties.js file?

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,208 @@
+              </div>
+            </div>
+          </div>
+        )}
```
Formatted according to tooltip component from Deck.gl: https://github.com/uber/deck.gl/blob/master/docs/developer-guide/interactivity.md#using-react

dhruv10 committed:
Small fixes like change name of functions, linting issues etc.
Moved Tooltip to a new component
Redirecting to NYC when the
 user denies location access using
a callback
Data sanity - used dataComparator functoin to not update incase of same data

dhruv10 committed:
1. Convert helper functions to arrow functions airbnb guidelines
2. Defined a function for initialControl Panel settings

dhruv10 committed:
Method getTooltipFormattedDetails implemented

dhruv10 committed:
Loading data based on
users response to location permission

prastut commented on src/VizPOIAnalytics/index.js:
```diff
@@ -0,0 +1,134 @@
+      FOAMTokenInUSD: await getValInUSD(),
+    });
+    this.setUserLocation();
+  }
```
Instead of setting state twice once for FOAMTokenInUSD and then setting user location,

1. fetch FOAM's value in USD
2. get userLocation from setUserLocation function (modified to getUserLocation)
3. When you have got approriate responses from 1 and 2, setState together.

dhruv10 replied:
Aren't they both different operations? Setting the value of FOAM token in getUserLocation() doesn't sounds like the function is doing job of another function.

prastut replied:
Check this code: https://es6console.com/k2rici3t/

prastut commented on src/VizPOIAnalytics/utils/helper.js:
```diff
@@ -0,0 +1,74 @@
+
+export const getTooltipFormattedDetails = async (allHoveredPOIDetails, FOAMTokenInUSD) => {
+  const sumOfFoamTokens = allHoveredPOIDetails.points
+    ? await getSumOfFoamTokens(allHoveredPOIDetails.points)
```
> await getSumOfFoamTokens

Why is there an await function here? getSumOfFOAMTokens is not a async function

dhruv10 replied:
Missed the fact. Resolved now.

dhruv10 committed:
Splitting POIAnalyticsControlPanel into components

dhruv10 committed:
Legend as a seperate component

dhruv10 committed:
1. Using axios
2. Sorting fetched data

dhruv10 committed:
small fix
